aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,570 @@
+
+                // Somewhat arbitrary yield points, to give the dispatcher a chance to signal
+                // another concurring future. This is mostly pertinent if we're running the test on
+                // a current_thread runtime. Otherwise, without await points, on a current_thread
+                // runtime, this function won't yield.
+                yield_now().await;
+                assert!(concurrency < self.max_concurrency);
+
+                let conn = conn.ok("hello");
+
+                yield_now().await;
+                assert!(concurrency < self.max_concurrency);
+
+                self.concurrency.fetch_sub(1, Ordering::Relaxed);
+                conn
+            }
+        }
+
+        runtime_flavor.build().block_on(async move {
+            let queue = Arc::new(LIFORequestQueue::new(concurrency, depth).unwrap());
+            let handler = Arc::new(queued_lifo(
+                Arc::clone(&queue),
+                ConcurrencyAssertingHandler {
+                    concurrency: Default::default(),
+                    max_concurrency: concurrency,
+                },
+            ));
+
+            join_all((0..requests).map(|_| {
+                let handler = Arc::clone(&handler);
+                async move {
+                    get("/").run_async(&handler).await;
+                }
+            }))
+            .await;
+        });
+    }
+
+    #[quickcheck]
+    fn quickcheck_lifo_cancel(parameters: Parameters) {
+        install_test_trace_subscriber();
+        debug!(?parameters, "quickcheck_lifo_cancel parameters");
+        let Parameters {
+            runtime_flavor,
+            depth,
+            concurrency,
+        } = parameters;
+
+        runtime_flavor.build().block_on(async move {
+            let unhang = Arc::new(Semaphore::new(0));
+            let queue = Arc::new(LIFORequestQueue::new(concurrency, depth).unwrap());
+            let handler = Arc::new(queued_lifo(
+                Arc::clone(&queue),
+                HangingHandler {
+                    unhang: Arc::clone(&unhang),
+                },
+            ));
+
+            let requests =
+                fill_queue(Arc::clone(&queue), Arc::clone(&handler), concurrency, depth).await;
+
+            // Abort all outstanding requests.
+            requests.iter().for_each(|request| request.abort());
+
+            // Wait until the queue is empty.
+            while queue.len() > 0 {
+                yield_now().await;
+            }
+
+            unhang.close();
+            let request = get("/").run_async(&handler).await;
+            assert_ok!(request);
+
+            for handle in requests {
+                // These handles will return a JoinError, but we're moreso interested to see if
+                // they've all terminated.
+                let _ = handle.await;
+            }
+        });
+    }
+
+    #[quickcheck]
+    fn quickcheck_lifo_full(parameters: Parameters) {
```

Comment:
Using quickcheck like this may be controversial... but I've uncovered some substantial bugs doing this.

Reply:
In hindsight something like https://github.com/madsim-rs/madsim?tab=readme-ov-file may have been a better fit, but I'm quite satisfied with how helpful quickcheck has been in automating permutations of outcomes of different scheduling of futures.